### DISCUSSION TITLE
How to solve the bottleneck on feature extraction

### DISCUSSION THREAD

**Discussion by bagustris** [felixbur/nkululeko#346] [General]
Currently, the bottleneck of nkululeko, afaik, is feature extraction (at least for the raw CSV dataset).  It means if we want different experiments (split, model) from the same features, we need to specify new experiment names. Imagine the following scenario:
- We want to change features from [FEATS] type = ['wavlm', 'hubert'] to type = ['hubert']
- We want to evaluate our model (testing) from database_a to database_b in one run
- We want to change training, dev, and test
- We want to test our model on a different database (previous test module?)

All scenario above requires new feature extraction and new experiment names. Using SSL, like HuBERT or WavLM, with big data, it takes hours to extract those features.

The following are possible ways to tackle limitation above: 
- Extract feature per dataset, not per split
- Update multidb to support multiple test sets (databases A, B, C for training, databases C, D for test)
- Main idea: same features, but different model (and parameter, etc)
- Shareable feature location (perhaps in `EXP` section by adding `store` location key)

**Comment by felixbur** [comment]
i would suggest to tackle this step by step, e.g. starting with extracting features per dataset, after the filters.
so
1) load datasets
2) apply filters (e.g. only 10 samples per speaker)
3) extract features per data set
4) combine them in the split implementation (i guess i would even model this with a new class: data-splitter)

**Reply by felixbur** [reply]
if you agree, I'd open an issue and start working on it, @bagustris 

perhaps after i did the [database.py simplifications](https://github.com/felixbur/nkululeko/issues/345)

**Reply by bagustris** [reply]
Go ahead, @felixbur. Some notes:
- Feature should be extracted once
- Per data set
- No matter the split, filter, etc.
- Maintain backward compatibility?

**Reply by felixbur** [reply]
>Maintain backward compatibility?

should work without having to change the config format, if that's what you mean

**Reply by bagustris** [reply]
Yes, the config format. It should work with the current example INI files in the `examples` folder. Of course, we need to run it again to get the new extracted features per dataset.

**Comment by bagustris** [comment]
I just asked Claude code, and here are several practical approaches for Transformer-based features, roughly ordered by impact:

## 1. Batch processing (biggest win)
Currently all extractors process one file at a time. Batching multiple files through the GPU in one forward pass gives 5–10× speedup.

The current loop:
```python
for file, start, end:
    signal = load(file)
    emb = model(signal)  # 1 file at a time
```
Should become:
```python
for batch in batches_of(files, batch_size=32):
    signals = [load(f) for f in batch]
    signals_padded = pad_sequence(signals)
    embs = model(signals_padded)  # 32 files at a time
```
This requires padding variable-length audio to the same length, then unpadding after.

## 2. `torch.compile()` (PyTorch 2.x)
One line added before the extraction loop:
```python
self.model = torch.compile(self.model)
```
Typically gives **20–40%** speedup on modern GPUs with no code changes.

## 3. Half precision (fp16/bf16)
```python
self.model = self.model.half()  # fp16
# then during inference:
y = torch.from_numpy(...).half().to(self.device)
```
Halves memory bandwidth, ~1.5–2× faster on Ampere+ GPUs.

## 4. `torch.no_grad()` + `torch.inference_mode()`
The code already uses `torch.no_grad()`. Switching to the stronger `torch.inference_mode()` avoids version counter overhead:
```python
with torch.inference_mode():
    ...
```

## 5. Parallel data loading with multiple workers
Pre-load audio from disk while GPU computes — use Python `ThreadPoolExecutor` or `multiprocessing` to load the next batch while the current one is being processed. For large datasets like yours (782k files), I/O can be a bottleneck.

## 6. Pre-resample / pre-convert audio files
The FLAC files causing errors are being decoded on-the-fly. Converting the entire dataset to 16kHz WAV once upfront eliminates per-file decoding overhead (FLAC decode is CPU-bound).

## 7. `audinterface.Feature` with `num_workers`
The audmodel/`auddim` extractors already use `audinterface` with `num_workers`. The SSL extractors could be refactored similarly — it handles batching and parallelism automatically.

---

**Practical recommendation for your setup**: The highest-ROI change with least code risk is **`torch.compile()`** (one line) + **`torch.inference_mode()`** (change from `no_grad`). Batching gives the most speedup but requires more refactoring of the extractor classes.

**Comment by felixbur** [comment]
>We want to change features from [FEATS] type = ['wavlm', 'hubert'] to type = ['hubert']

this should actually not be a problem with the original design

**Reply by felixbur** [reply]
yes: tried it and it works, at least with [os, praat] vs [os]
